Feature/kas 3206 refactor van modal om nieuwe zitting aan te maken naar octane

sergiofenoll opened the pull request:
Jira: https://kanselarij.atlassian.net/jira/software/c/projects/KAS/boards/1?modal=detail&selectedIssue=KAS-3206&assignee=621cf63114cd24006906e31c

Renames the component & refactors for Octane syntax. Replaces as many Vl-components to Auk-components as possible (Datepicker still TODO because `Auk::Datepicker` does not support selecting a time).

Also fixes a bug in the `generateNewNumber()` method where it would break if there were too many future agendas.

Note: didn't run tests (yet), but adapted the selectors to deal with the change from `Webcomponent::VlModalFooter` to `Auk::Modal::Footer`.

sergiofenoll committed:
Refactor New Session Form

Renamed from Sessions::Forms::NewSession to Meeting::NewMeetingModal

Using Octane now, and replaced as many Vl-components with Auk components
as possible.

sergiofenoll committed:
Fix logic for retrieving newest meeting number

Filter on meetings from this year & get the highest number, all in
mu-cl-resources. Otherwise we risk missing meetings with higher numbers
if there's more than 20 future meetings.

sergiofenoll committed:
Add data-test- attribute for adapted save button

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+      `'onCancel' is required and must be a function`,
+      this.args.onCancel !== undefined &&
+        typeof this.args.onCancel === 'function'
+    );
```
Currently we don't have these kind of checks in components, so I would leave them out for now. What we try do to more consistently is adding some documentation about the expected arguments. E.g. https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/components/publications/publication/request-activity-panel.js#L6-L11  You can provide some additional information in that comment block in case the argument name isn't obvious.

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+        (meeting) =>
+          meeting.plannedStart &&
+          meeting.plannedStart.getFullYear() === this.currentYear
+      );
```
Can't we add this as an additional filter on the backend query?  Something like `':lte:planned-start': last-day-of-the-year`.

If so, given we sort on `-number`, we only need 1 result. You can use [our custom story extension `this.queryOne`](https://github.com/kanselarij-vlaanderen/frontend-kaleidos/blob/development/app/services/store.js#L7) for that.

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+    return isAnnexMeetingKind(this.selectedKindUri);
+  }
+
+  async generateNewNumber() {
```
Can you convert this to a concurrency-task and disable the meeting number input-field as long as this task is running.
Maybe `initializeMeetingNumber` is a more clear name for the method?

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+        .map((meeting) => meeting.number)
+        .filter((meetingId) => meetingId !== undefined);
+      let id = 0;
+      // FIX voor de eerste agenda van het jaar -> Anders math.max infinity
```
You can remove this comment. It confuses me more than just reading the code :smiley:

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+  @service agendaService;
+  @service newsletterService;
+  @service toaster;
+  @service formatter;
```
Can you replace usages of this service with inline date formatting? We should get rid of this service in the long run, since it doesn't have an added value, but just adds complexity to the code.
The case `this.formatter.formatDate(null)` can even just be replaced with `new Date()`.

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+      createdFor: meeting,
+      status,
+      created: date || fallBackDate,
+      modified: date || fallBackDate,
```
As far as I see, `date` is always present in the current code flow. Fallback can be removed.

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+      modified: date || fallBackDate,
+    });
+    const savedAgenda = await agenda.save();
+    return savedAgenda;
```
To be consistent with most of the codebase, I would replace with
```javascript
await agenda.save();
return agenda;
```

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+  async createAgendaitemToApproveMinutes(agenda, newMeeting, closestMeeting) {
+    const now = new Date();
+
+    // load code-list item
```
Remove comment. Code is clear imo.

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+      CONSTANTS.DECISION_RESULT_CODE_URIS.GOEDGEKEURD
+    );
+
+    // Treatment of agenda-item / decision activity
```
Remove comment. Code is clear imo.

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+      isApproval: true,
+      treatments: A([agendaItemTreatment]),
+    });
+    return await agendaitem.save();
```
To be consistent with most of the codebase, I would replace with
```javascript
await agendaitem.save();
return agendaitem;
```

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+      formallyOk: CONSTANTS.ACCEPTANCE_STATUSSES.NOT_YET_OK,
+      mandatees: [],
+      pieces: [],
+      approvals: [],
```
No need to initialize those empty relations

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+    return await agendaitem.save();
+  }
+
+  @task({ drop: true })
```
Fyi, there is also `@dropTask` as an alias

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+  }
+
+  @task({ drop: true })
+  *createNewSession() {
```
Rename to `createMeeting`

erikap replied:
Can you rearrange the order of the methods such that `createMeeting` is above the 'helper' methods `createAgenda` and `createAgendaitemToApproveMinutes`?

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+    });
+
+    const closestMeeting =
+      yield this.agendaService.getClosestMeetingAndAgendaId(startDate);
```
This is the only place where `getClosestMeetingAndAgendaId(startDate)` is used in the codebase. It just does a custom call to the backend and is stateless. Can you convert it to a util (or maybe add to existing `meeting-utils.js`) and import that one instead of using the `agendaService`.  I also have a preference to name it `fetch...` instead of `get...` when it includes a call to the backend.

Same reasoning holds for `agendaService.assignNewSessionNumbers`.

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+    const kindUriToAdd = selectedKindUri || CONFIG.MINISTERRAAD_TYPES.DEFAULT;
+    const date = this.formatter.formatDate(null);
+    const startDate = this.startDate || date;
+    const newMeeting = this.store.createRecord('meeting', {
```
Just `meeting` instead of `newMeeting`?

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+      meetingNumber,
+      formattedMeetingIdentifier,
+    } = this;
+    const kindUriToAdd = selectedKindUri || CONFIG.MINISTERRAAD_TYPES.DEFAULT;
```
Fallback not needed, since dropdown to select kind always has a value (and there is no option to clear)

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+  }
+
+  @action
+  meetingNumberChangedAction(meetingNumber) {
```
Use an active action names instead of passive one. E.g. `updateMeetingIdentifier`

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+  @action
+  meetingNumberChangedAction(meetingNumber) {
+    this.meetingNumber = meetingNumber;
+    this.formattedMeetingIdentifier = `VR PV ${this.currentYear}/${meetingNumber}`;
```
This formatted string is constructed at several locations in this component. It may be better to extract it to a separate method which sets the formatted identifier (a.o. based on `this.isAnnexMeeting`)

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+  }
+
+  @action
+  saveAction() {
```
Name is too generic. It's not clear what the action does. Action can maybe even be merged with the method from the previous comment. If a value is given as argument, it is assigned to `this.formattedMeetingIdentifier`; otherwise the formatted identifier is automatically constructed.

erikap commented on app/components/meeting/new-meeting-modal.js:
```diff
@@ -0,0 +1,226 @@
+  @task({ drop: true })
+  *createNewSession() {
+    const {
+      isDigital,
```
Doesn't seem to be used and is not defined on the `meeting` model. Can be removed.

erikap commented:
> Renames the component & refactors for Octane syntax. Replaces as many Vl-components to Auk-components as possible (Datepicker still TODO because `Auk::Datepicker` does not support selecting a time).

@brenner-company Staat het al op je roadmap om `Auk::Datepicker` uit te breiden met support voor time selection?

brenner-company commented:
@erikap Dit staat inderdaad al op de roadmap en zou volgende sprint kunnen opgenomen worden 👍

Momenteel ben ik binnen https://kanselarij.atlassian.net/browse/KAS-3182 bezig met het syncen van het design (Mono) met de datepickers. Ben daar quasi klaar mee (enkel time input heeft nog beetje werk nodig):

<img width="388" alt="CleanShot 2022-03-04 at 10 02 42@2x" src="https://user-images.githubusercontent.com/18174827/156732831-aab6b92d-3dc2-490d-a56a-9e5b8be98877.png">

erikap commented:
:+1: Dan laat ik deze PR niet blokkeren op het gebruik van `WebComponents::VlDatepicker`. Er staat een TODO bij om in de toekomst te replacen met `Auk::Datepicker`.

sergiofenoll committed:
Utils::KindSelector propagates default to parent

If no selectedKind is passed to Utils::KindSelector, the first "kind" is
used to fill the dropdown. The previous behaviour didn't "set" said default
in the parent, this is now fixed by calling the `setAction` action in the constructor.

sergiofenoll committed:
Apply review feedback

sergiofenoll commented:
> This formatted string is constructed at several locations in this component. It may be better to extract it to a separate method which sets the formatted identifier (a.o. based on this.isAnnexMeeting)

> Name is too generic. It's not clear what the action does. Action can maybe even be merged with the method from the previous comment. If a value is given as argument, it is assigned to this.formattedMeetingIdentifier; otherwise the formatted identifier is automatically constructed.

@erikap resolved these two by wrapping `meetingNumber` and `formattedMeetingIdentifier` in getters & setters. I realize now that it may be a little confusing because it's hiding behaviour behind assignments, so if that's undesired let me know and I'll refactor it into a regular method and add calls where necessary.

erikap committed:
Inline destructuring of this in new-meeting-modal component

Currenty a mix of destructuring and property access are used. Converting
everything to direct property access on this.

sergiofenoll committed:
Remove unnecessary updateMeetingNumber action

Since the meetingNumber is bound to the input, the action is no longer
needed. Also use the @width argument of Auk::Input instead of passing in
the full class

erikap committed:
Fix test selector for meeting kind selector

erikap committed:
Add missing test selector for main meeting dropdown

erikap reviewed (changes requested):
The current refactor of `KindSelector` component is complex with the get/set construction and breaks the preselected value in the `EditSession` component. This can imo only be cleanly fixed by refactoring the `EditSession` component as well.
There is also a future story to refactor the `KindSelector` component to retrieve the codelist from the backend. So, for this PR, I would revert the `KindSelector` component changes such that this PR can pass.

sergiofenoll committed:
Revert "Utils::KindSelector propagates default to parent"

This reverts commit 78e81e982bbeec6abe62be01eb166784d98aea1d.

sergiofenoll committed:
Set fallback kindUri in case KindSelector hasn't been interacted with
